File: logics/traits/factory.rs
use ink::prelude::vec::Vec;
use openbrush::traits::AccountId;

use super::registry;

#[openbrush::wrapper]
pub type FactoryRef = dyn Factory;

#[openbrush::trait_definition]
pub trait Factory {
    #[ink(message)]
    fn create(&self, asset: AccountId, data: Vec<u8>) -> Result<AccountId>;
}

#[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
pub enum Error {
    PoolImplementationMissing,
    CallerIsNotManager,
    Registry(registry::Error),
    Custom(u8),
}

pub type Result<T> = core::result::Result<T, Error>;

File: frontend/src/__generated__/types/types-arguments/factory.ts
import type BN from 'bn.js';

export type AccountId = string | number[]

export type Hash = string | number[]

export enum LangError {
	couldNotReadInput = 'CouldNotReadInput'
}

export interface factory_Error {
	poolImplementationMissing ? : null,
	callerIsNotManager ? : null,
	registry ? : registry_Error,
	custom ? : (number | string | BN)
}

export class factory_ErrorBuilder {
	static PoolImplementationMissing(): factory_Error {
		return {
			poolImplementationMissing: null,
		};
	}
	static CallerIsNotManager(): factory_Error {
		return {
			callerIsNotManager: null,
		};
	}
	static Registry(value: registry_Error): factory_Error {
		return {
			registry: value,
		};
	}
	static Custom(value: (number | string | BN)): factory_Error {
		return {
			custom: value,
		};
	}
}

export enum registry_Error {
	poolAlreadyExists = 'PoolAlreadyExists',
	callerIsNotManager = 'CallerIsNotManager'
}

